Adding bicep resource type

davidfowl opened the pull request:
- [x] - Define custom bicep template resource type in Aspire.Hosting.Azure
    - [x] Added manifest support
- [x] - Add support for Aspire.Hosting.Azure.Provisioning
    - [x] - Move to rest API calls instead of shelling to az (still shelling to az for bicep)
- [x] - Added parallel bicep-based resources for existing resources
    - [x] - Storage
    - [x] - KeyVault
    - [x] - CosmosDb
    - [x] - Azure SQL
    - [x] - AppConfiguration
    - [x] - AppInsights
    - [x] - Redis
    - [x] - ServiceBus
- [x] - Added support for Azure Postgres
- [x] - Write logs of bicep (mostly lifted from azd) 
- [x] - Add emulator support 
- [x] - Add doc comments
- [x] - Add basic tests

Fixes https://github.com/dotnet/aspire/issues/1941

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/2056)

mitchdenny committed:
Added bicep resource primitive for azure resources
- Added new resource types to aid in the transition to the new bicep resource primitive
- Updated a BicepProvisioner to AzureProvisioning
- Added lots of bicep files

mitchdenny commented on playground/azuresql/AzureSql.AppHost/Program.cs:
```diff
@@ -0,0 +1,37 @@
+
+           var hostName = db.Resource.Output["sqlServerName"];
+           
+           var cs = $"Server=tcp:{hostName}.database.windows.net,1433;Initial Catalog={databaseName};Encrypt=True;TrustServerCertificate=False;Connection Timeout=30;Authentication=\"Active Directory Default\";";
```
You really want the entire connection string to come from the resource provider. The reason is that the above code only works on Azure Public cloud. If you are using something like mooncake or one of the other non-public Azure clouds the endpoints for the resources will be different. I realize this is code that the end developer writes, but we would probably wrap AddBicepTemplate with more concrete resource types.

davidfowl replied:
You mean the host name, there’s no connection string property that comes from SQL.

This is a primitive, and an escape hatch 🤓

mitchdenny commented on src/Aspire.Hosting.Azure.Provisioning/Provisioners/BicepProvisioner.cs:
```diff
@@ -0,0 +1,148 @@
+            return false;
+        }
+
+        foreach (var item in section.GetSection("Outputs").GetChildren())
```
What happens if you change the template? You might need to store a hash of the bicep file here somewhere so you can do change detection, and if that occurs execute the deployment again (you would need to nuke all the output values in that case).

mitchdenny commented on src/Aspire.Hosting.Azure/AzureBicepResource.cs:
```diff
@@ -0,0 +1,15 @@
+
+public class AzureBicepResource(string name, string templateFile) : Resource(name), IAzureResource
+{
+    public string TemplateFile { get; } = templateFile;
```
This will need to be more ellaborate. There are some deployment scenarios where you need to be able to use modules/sub deployments to solve a problem. A common example is when you are dealing wtih role assignments and/or tags. If you want to _add a tag_ without destructively removing ones you need a sub deployment. To work around some limitations with the role assignment resource you'll sometimes need a sub module.

I think the way to do this is that we have a method that we pass in a BicepEmissionContext (or something) where you can attach files to it (with a primary entrypoint for a specific resource). It would be the job of the manifest writer (or provisioner in this case) to write those files to disk.

davidfowl replied:
An example would be helpful.

mitchdenny commented on src/Aspire.Hosting.Azure/AzureResourceExtensions.cs:
```diff
@@ -15,6 +16,56 @@ namespace Aspire.Hosting;
+        return builder;
+    }
+
+    public static IResourceBuilder<AzureBicepResource> WithParameter(this IResourceBuilder<AzureBicepResource> builder, string name, IResourceBuilder<IResourceWithConnectionString> value)
```
Should this be a special WithReference for consistency?

davidfowl replied:
With isn’t a reference. I’ll call it AddParameter

mitchdenny commented on src/Aspire.Hosting/Dcp/Process/ProcessUtil.cs:
```diff
@@ -95,7 +95,7 @@ public static (Task<ProcessResult>, IAsyncDisposable) Run(ProcessSpec processSpe
         try
         {
-            AspireEventSource.Instance.ProcessLaunchStart(processSpec.ExecutablePath, processSpec.Arguments ?? "");
+            // AspireEventSource.Instance.ProcessLaunchStart(processSpec.ExecutablePath, processSpec.Arguments ?? "");
```
Why?

davidfowl replied:
Shared code, I didn’t attempt to pull it over into azure provisioning

mitchdenny commented on playground/azuresql/AzureSql.AppHost/Program.cs:
```diff
@@ -0,0 +1,37 @@
+
+           if (ctx.PublisherName == "manifest")
+           {
+               ctx.EnvironmentVariables["ConnectionStrings__db"] = $"Server=tcp:{{{db.Resource.Name}.outputs.sqlServerName}}.database.windows.net,1433;Initial Catalog={databaseName};Encrypt=True;TrustServerCertificate=False;Connection Timeout=30;Authentication=\"Active Directory Default\";";
```
We need to think about how we use Bicep outputs. In ARM deployments, outputs can't be flagged as secrets. For resource types that support managed identity auth that isn't such an issue because the connection string doesn't contain any real secrets, but it can be an issue for those resource types that don't support managed auth.

davidfowl replied:
Are you talking about this low level primitive or something else (higher level resource types)?

davidfowl committed:
Added app insights bicep resource

davidfowl committed:
Small resource clean up

davidfowl committed:
Blow up if outputs aren't set

davidfowl committed:
Added support for azure redis

davidfowl committed:
Added redis to sample

davidfowl committed:
Added blob endpoint

davidfowl committed:
Added table endpoint

mitchdenny commented on playground/bicep/BicepSample.AppHost/aspire-manifest.json:
```diff
@@ -0,0 +1,181 @@
+    },
+    "cosmos": {
+      "type": "azure.bicep.v0",
+      "connectionString": "AccountEndpoint={cosmos.outputs.documentEndpoint};AccountKey={keys({cosmos.outputs.accountName})}",
```
Is this something that AZD needs to understand so it can go fetch a key from the resource?

davidfowl replied:
Yep

davidfowl committed:
Added service bus
- Added storage queue and service bus example

davidfowl committed:
Gen new manifest

davidfowl committed:
Put the full resource name in the keys method

davidfowl committed:
Fixed formatting

davidfowl committed:
Tweaks and review feedback
- Renamed AddParameter to WithParameter
- Make ProcessUtl more sharable, still needs to be moved.
- Get deployment outputs using the API instead of shelling out

davidfowl committed:
Deploy using the arm APIs instead

davidfowl committed:
Oops

davidfowl committed:
Added emulator support back

davidfowl committed:
The only relevant context is the user

davidfowl committed:
WIP docs

davidfowl committed:
Fixed errors

davidfowl committed:
Fixed other errors

davidfowl committed:
More doc comments

davidfowl committed:
Clean up

mitchdenny commented on playground/bicep/BicepSample.AppHost/Program.cs:
```diff
@@ -0,0 +1,65 @@
+var user = builder.AddParameter("username");
+var pwd = builder.AddParameter("password", secret: true);
+
+var pg = builder.AddBicepAzurePostgres("postgres2", user, pwd).AddDatabase("db2");
```
I think for something like this I would be inclined to let the AddBicepAzurePostgres do the wireup to automatically generate the password (using the `"input":` field mechanism).

I think using a parameter is probably more if you want to use an extenral connection string (in this particular case).

mitchdenny commented on playground/bicep/BicepSample.AppHost/Program.cs:
```diff
@@ -0,0 +1,65 @@
+                   .WithParameter("test", parameter)
+                   .WithParameter("values", ["one", "two"]);
+
+var kv = builder.AddBicepKeyVault("kv");
```
For this Azure specific resource types is the plan to completely replace the existing keyvault resource types (and rename these new ones). The fact that they are BicepResource types shouldn't be a factor right?

davidfowl replied:
Yep

davidfowl committed:
Added unit tests

davidfowl committed:
Small tweaks

mitchdenny reviewed (approved):
Approving after a detailed walkthrough from David. Would be good to get this in and keep iterating.
